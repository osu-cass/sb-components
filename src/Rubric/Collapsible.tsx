import * as React from "react";

export interface CollapsibleProps {
  className?: string;
  style?: React.CSSProperties;
  label: string;
}

export interface CollapsibleState {
  isCollapsed: boolean;
}

export class Collapsible extends React.Component<
  CollapsibleProps,
  CollapsibleState
> {
  constructor(props: CollapsibleProps) {
    super(props);
    this.state = {
      isCollapsed: true
    };
  }

  toggleCollapsed() {
    this.setState({
      isCollapsed: !this.state.isCollapsed
    });
  }

  render() {
    const label = (this.state.isCollapsed ? "▶ " : "▼ ") + this.props.label;
    const style: React.CSSProperties = this.state.isCollapsed
      ? { display: "none" }
      : {};

    return (
      <div
        className={this.props.className}
        style={this.props.style}
        aria-expanded={!this.state.isCollapsed}
      >
        <a
        role="button"
          className="collapsible-label link-button"
          href="javascript:"
          onClick={() => this.toggleCollapsed()}
          tabIndex={0}
        >
          {label}
        </a>
        <div
          className="collapsible-body"
          style={style}
          aria-hidden={this.state.isCollapsed}
        >
          {this.props.children}
        </div>
      </div>
    );
  }
}
