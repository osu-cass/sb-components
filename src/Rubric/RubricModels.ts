export interface SampleResponseModel {
  purpose: string;
  scorePoint: string;
  name: string;
  sampleContent: string;
}

export interface RubricSampleModel {
  maxValue: string;
  minValue: string;
  sampleResponses: SampleResponseModel[];
}

export interface RubricEntryModel {
  scorepoint: string;
  name: string;
  value: string;
}

export interface RubricModel {
  language: string;
  rubricEntries: RubricEntryModel[];
  samples: RubricSampleModel[];
}

export interface SampleItemScoringModel {
  answerKey?: string;
  hasMachineRubric?: boolean;
  scoringOptions?: SmarterAppOptionModel[];
  rubrics?: RubricModel[];
}

export interface SmarterAppOptionModel {
  name: string;
  value: string;
  feedback: string;
  language: string;
  answer: boolean;
}

export interface RubricTableRowModel {
  score: string;
  rationale: string;
  sample: string;
}
