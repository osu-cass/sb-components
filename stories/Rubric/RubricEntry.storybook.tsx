import * as React from "react";
import { storiesOf } from "@storybook/react";
import { RubricEntry, RubricEntryModel } from "@src/index";
import { centerDecorator } from "../CenterDecorator";

const rubricEntryMockProps: RubricEntryModel = {
  scorepoint: "",
  name: "\n        Rubric ",
  value:
    '<p style="color:#000000; ">Exemplar:</p><p style="color:#000000; ">&#xA0;</p><p style="color:#000000; ">4; 3; 1, 2, 4 OR 9; 3; 1, 3, 9</p><p style="color:#000000; ">&#xA0;</p><p style="color:#000000; ">For this item, a full-credit response includes</p><p style="color:#000000; ">&#xA0;</p><ul style="padding-left:1.5000em; "><li><p style="color:#000000; background-color:transparent; ">1 point: Student gives one of these two combinations of answers:</p><p style="color:#000000; background-color:transparent; ">&#xA0;</p></li></ul><div align="margins"><table class="newtable" id="item_3322_Table1" style="width:15.241799999999997em; border-spacing:0;margin-left:0.5069in; margin-right:4.3368in; " cellpadding="0" cellspacing="0"><colgroup style="width:4.8792em; " /><colgroup style="width:5.4792em; " /><colgroup style="width:4.8792em; " /><tbody><tr><td style="padding:0.2292em;  border-left:0.06em solid #000000; border-top:0.06em solid #000000; border-bottom:0.06em solid #000000;vertical-align:top; "><p style="color:#000000; font-weight:bold; font-size:1.0000em; font-family:Verdana; ">Part A </p><p style="color:#000000; font-weight:bold; font-size:1.0000em; font-family:Verdana; ">First Box</p></td><td style="padding:0.2292em;  border-left:0.06em solid #000000; border-top:0.06em solid #000000; border-bottom:0.06em solid #000000;vertical-align:top; "><p style="color:#000000; font-weight:bold; font-size:1.0000em; font-family:Verdana; ">Part A </p><p style="color:#000000; font-weight:bold; font-size:1.0000em; font-family:Verdana; ">Second Box</p></td><td style="padding:0.2292em;  border:0.06em solid #000000;vertical-align:top; "><p style="color:#000000; font-weight:bold; font-size:1.0000em; font-family:Verdana; ">Part B</p></td></tr><tr><td style="padding:0.2292em;  border-left:0.06em solid #000000; border-bottom:0.06em solid #000000;vertical-align:top; "><p style="color:#000000; font-size:1.0000em; font-family:Verdana; ">4</p></td><td style="padding:0.2292em;  border-left:0.06em solid #000000; border-bottom:0.06em solid #000000;vertical-align:top; "><p style="color:#000000; font-size:1.0000em; font-family:Verdana; ">3</p></td><td style="padding:0.2292em;  border-left:0.06em solid #000000; border-right:0.06em solid #000000; border-bottom:0.06em solid #000000;vertical-align:top; "><p style="color:#000000; font-size:1.0000em; font-family:Verdana; ">1, 2, 4</p></td></tr><tr><td style="padding:0.2292em;  border-left:0.06em solid #000000; border-bottom:0.06em solid #000000;vertical-align:top; "><p style="color:#000000; font-size:1.0000em; font-family:Verdana; ">9</p></td><td style="padding:0.2292em;  border-left:0.06em solid #000000; border-bottom:0.06em solid #000000;vertical-align:top; "><p style="color:#000000; font-size:1.0000em; font-family:Verdana; ">3</p></td><td style="padding:0.2292em;  border-left:0.06em solid #000000; border-right:0.06em solid #000000; border-bottom:0.06em solid #000000;vertical-align:top; "><p style="color:#000000; font-size:1.0000em; font-family:Verdana; ">1, 3, 9</p></td></tr></tbody></table></div><ul style="padding-left:1.5000em; "><li><p style="color:#000000; ">&#xA0;</p><p style="color:#000000; ">(1 point)</p><p style="color:#000000; ">&#xA0;</p></li><li><p style="color:#000000; ">0 points: Any other response. (0 points).</p></li></ul>'
};

storiesOf("Rubric Entry", module)
  .addDecorator(centerDecorator)
  .add("with an empty entry", () => (
    <RubricEntry scorepoint="" name="" value="" />
  ))
  .add("with a rubric entry", () => <RubricEntry {...rubricEntryMockProps} />);
