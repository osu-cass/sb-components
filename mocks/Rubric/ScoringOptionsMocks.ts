import { SmarterAppOptionModel } from "./RubricModels";

export const scoringOptionsMock: SmarterAppOptionModel[] = [
  {
    name: "A",
    value: "14",
    feedback: "CORRECT: Because its the right answer",
    language: "EN-US",
    answer: true
  },
  {
    name: "B",
    value: "a million",
    feedback: "INCORRECT: Because its the wrong answer",
    language: "EN-US",
    answer: false
  }
];
