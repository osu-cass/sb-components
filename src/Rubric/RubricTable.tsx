import * as React from "react";
import { RubricModel, RubricTableRowModel } from "./RubricModels";
import { ItemRevisionModel } from "@src/index";

export interface RubricTableProps {
  rubrics: RubricModel[];
  itemModel?: ItemRevisionModel;
}

export class RubricTable extends React.Component<RubricTableProps, {}> {
  renderRubric(rubric: RubricModel, index: number) {
    let rows: RubricTableRowModel[] = [];
    if (rubric.rubricEntries && rubric.rubricEntries.length > 0) {
      rows = rubric.rubricEntries.map(entry => {
        const sample = rubric.samples.find(
          s =>
            s.sampleResponses &&
            s.sampleResponses[0] &&
            s.sampleResponses[0].scorePoint === entry.scorepoint
        );
        const sampleHtml = sample
          ? sample.sampleResponses.map(sr => sr.sampleContent).join("<br/>")
          : "";

        return {
          score: entry.scorepoint,
          rationale: this.populatePath(entry.value),
          sample: sampleHtml
        };
      });
    }

    const showSample = rubric.samples.length !== 0;

    const leftAlign: React.CSSProperties = {
      textAlign: "left"
    };

    // tslint:disable:react-no-dangerous-html
    const rowsJsx = rows.map((row, i) => (
      <tr key={i}>
        <td>{row.score}</td>
        <td
          dangerouslySetInnerHTML={{ __html: row.rationale }}
          style={leftAlign}
        />
        {showSample ? (
          <td
            dangerouslySetInnerHTML={{ __html: row.sample }}
            style={leftAlign}
          />
        ) : (
          undefined
        )}
      </tr>
    ));
    // tslint:enable:react-no-dangerous-html

    return (
      <table className="item-data-table" key={index}>
        <thead>
          <tr>
            <th>Score</th>
            <th>Rationale</th>
            {showSample ? <th>Exemplar</th> : undefined}
          </tr>
        </thead>
        <tbody>{rowsJsx}</tbody>
      </table>
    );
  }

  populatePath(html: string) {
    const { itemModel } = this.props;
    const rubricElement = document.createElement("div");
    // tslint:disable-next-line
    rubricElement.innerHTML = html;
    const imgTags = rubricElement.getElementsByTagName("img");
    Array.from(imgTags).forEach(img => {
      const pos = img.src.indexOf("item_");
      if (pos && itemModel) {
        const text = img.src;
        const path = "rubricImage/";
        let params = `?itemId=${itemModel.bankKey}-${itemModel.itemKey}`;
        params += itemModel.revision ? `&version=${itemModel.revision}` : "";
        img.src = [text.slice(0, pos), path, text.slice(pos), params].join("");
      }
    });

    return rubricElement.innerHTML;
  }

  render() {
    return (
      <div>{this.props.rubrics.map((r, i) => this.renderRubric(r, i))}</div>
    );
  }
}
