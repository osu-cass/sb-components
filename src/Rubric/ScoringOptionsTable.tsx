import * as React from "react";
import { SmarterAppOptionModel } from "./RubricModels";

export interface ScoringOptionsProps {
  options: SmarterAppOptionModel[];
}

// tslint:disable-next-line:variable-name
export const ScoringOptions: React.SFC<ScoringOptionsProps> = props => {
  // tslint:disable:react-no-dangerous-html
  const rowsJSX = props.options.map(o => (
    <tr>
      <td>{o.name}</td>
      <td dangerouslySetInnerHTML={{ __html: o.feedback }} />
    </tr>
  ));
  // tslint:enable:react-no-dangerous-html

  return (
    <table className="item-data-table">
      <thead>
        <tr>
          <th>Option</th>
          <th>Feedback</th>
        </tr>
      </thead>
      <tbody>{rowsJSX}</tbody>
    </table>
  );
};
