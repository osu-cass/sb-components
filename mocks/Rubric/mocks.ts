import { RubricModalProps } from "@src/Rubric/RubricModal";

export const rubricModalMockProps: RubricModalProps = {
  showModal: true,
  rubrics: [
    {
      language: "English",
      rubricEntries: [
        {
          scorepoint: "2",
          name: "\n        Rubric 2",
          value:
            '<p style="text-decoration:underline; font-weight:bold; ">3 point text </p><p style="">&#xA0;</p><p style="">The student determines that Emily can make it to school on time at her current rate of speed and includes a valid explanation containing a full chain of reasoning that supports this conclusion. The student may make minor computation errors that do not affect the reasonableness of the explanation. </p><p style="">&#xA0;</p><p style="">&#xA0;</p><p style="text-decoration:underline; font-weight:bold; ">2 point text </p><p style="text-decoration:underline; font-weight:bold; ">&#xA0;</p><p style="">The student determines that Emily can make it to school on time at her current rate of speed and includes a valid explanation containing an incomplete chain of reasoning that supports this conclusion. (An incomplete chain of reasoning can be defined by missing process steps or unsupported calculations in an otherwise complete chain of reasoning.) </p><p style="">OR </p><p style="">The student determines that Emily can make it to school on time at her current rate of speed and includes a valid explanation containing a full chain of reasoning that supports this conclusion, but makes computation errors that affect the reasonableness of the explanation. </p>'
        },
        {
          scorepoint: "1",
          name: "\n        Rubric 1",
          value:
            '<p style="">The student completes the task and reaches a conclusion. The student\'s explanation attempts to relate distance to time, but contains errors in fundamental mathematical procedures. </p><p style="">&#xA0;</p>'
        },
        {
          scorepoint: "0",
          name: "\n        Rubric 0",
          value:
            '<p style="">The student demonstrates a lack of comprehension in regard to the mathematical content and practices essential to the task. </p>'
        }
      ],
      samples: [
        {
          maxValue: "2",
          minValue: "2",
          sampleResponses: [
            {
              purpose: "OtherExemplar",
              scorePoint: "2",
              name: "2-Point Other Official Sample Answers\n      ",
              sampleContent:
                '<p style=""><span style="text-decoration:underline; font-weight:bold; ">Sample 1: </span>&#xA0;</p><p style="">Emily can travel 3/4 mile in under 5 minutes, so she can travel 3 miles (4x as far) in under 20 minutes (4x as long.) This means that Emily will have traveled 3 miles before 8:45 (20 minutes after she left.) Then there is only .42 miles left to go, and since .42 miles is less than 3/4 mile, we know she can cover that distance in less than 5 minutes. That means she will get to school before 8:50, so she will be on time. </p><p style="">&#xA0;</p><p style=""><span style="text-decoration:underline; font-weight:bold; ">Sample 2:</span> &#xA0;</p><p style="">Emily travels 3/4 mile in 4.5 minutes, so to find her rate of travel we would divide time by distance and get (4.5) / (.75) = 6 minutes per mile. Multiply 6 minutes per mile times the distance she has to travel (3.42 miles) to find the time it would take for her to get to school (20.52.) Since 20.52 is less than 21 minutes, we know it will take her less than 21 minutes to get to school. 8:25 + 21 minutes is 8:46, which is before 8:50, so she will make it to school on time. </p>'
            }
          ]
        }
      ]
    }
  ]
};
