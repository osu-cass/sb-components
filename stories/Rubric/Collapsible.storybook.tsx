import * as React from "react";
import { storiesOf } from "@storybook/react";
import { Collapsible } from "../../src/Rubric/Collapsible";

const style = {
  height: "300px",
  width: "300px",
  background: "lightgrey"
}

const collapsibleStyle = {
  border: "1px solid red"
}

const box = ( <div style={style} >Collapsible Content</div> );

storiesOf( "Collapsible", module )
  .add( "No classname of style", () => (
    <Collapsible className="" style={{}} label="label">{box}</Collapsible>
  ) )
  .add( "Collapsible with inline red border", () => (
    <Collapsible className="" style={{...collapsibleStyle}} label="label">{box}</Collapsible>
  ) )
  .add( "Collapsible with CSS classname", () => (
    <Collapsible className="test-collapsible" style={{}} label="label">{box}</Collapsible>
  ) )
