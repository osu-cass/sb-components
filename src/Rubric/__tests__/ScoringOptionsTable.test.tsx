import * as React from "react";
import * as ReactDOM from "react-dom";
import { shallow } from "enzyme";
import { SmarterAppOptionModel } from "../RubricModels";
import { scoringOptionsMock } from "@mocks/Rubric/ScoringOptionsMocks";
import { ScoringOptions } from "../ScoringOptionsTable";

describe("ScoringOptionsTable", () => {
  it("renders correctly", () => {
    const wrapper = shallow(<ScoringOptions options={scoringOptionsMock} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("renders correctly", () => {
    const wrapper = shallow(<ScoringOptions options={[]} />);
    expect(wrapper).toMatchSnapshot();
  });
});
