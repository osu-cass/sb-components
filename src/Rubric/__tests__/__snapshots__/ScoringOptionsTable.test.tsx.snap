// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScoringOptionsTable renders correctly 1`] = `
<table
  className="item-data-table"
>
  <thead>
    <tr>
      <th>
        Option
      </th>
      <th>
        Feedback
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        A
      </td>
      <td
        dangerouslySetInnerHTML={
          Object {
            "__html": "CORRECT: Because its the right answer",
          }
        }
      />
    </tr>
    <tr>
      <td>
        B
      </td>
      <td
        dangerouslySetInnerHTML={
          Object {
            "__html": "INCORRECT: Because its the wrong answer",
          }
        }
      />
    </tr>
  </tbody>
</table>
`;

exports[`ScoringOptionsTable renders correctly 2`] = `
<table
  className="item-data-table"
>
  <thead>
    <tr>
      <th>
        Option
      </th>
      <th>
        Feedback
      </th>
    </tr>
  </thead>
  <tbody />
</table>
`;
